src/component.js

```diff
@@ -53,13 +55,12 @@ extend(Component.prototype, {
 	 * 	updated
 	 */
 	setState(state, callback) {
-		if (!this.prevState) this.prevState = this.state;
-		this.state = extend(
-			extend({}, this.state),
-			typeof state === 'function' ? state(this.state, this.props) : state
-		);
 		if (callback) this._renderCallbacks.push(callback);
-		enqueueRender(this);
+		this._nextState = extend(
+			extend({},this._nextState),
+			typeof state==='function' ? state(this._nextState, this.props) : state
+		);
+		enqueueRender(this, this.props, this.context);
```

Comment:
I believe this breaks when enqueuing multiple setState() updates? Struggling to think of a way to demonstrate but it seems like the value of `this.props` being pushed into the render queue will result in it being incorrect for intermixed prop+state driven updates to the same component.

Reply:
I would be surprised if what you describe is not already covered by current tests. I understand the concern, especially with the recent regression involving the mutable setState fix, in issue https://github.com/developit/preact/issues/1168 and https://github.com/developit/preact/issues/1170

(Had I made a PR instead of just commenting that regression might have been avoided.)

I hope you see that these changes will make similar regressions less likely to happen because there will be less conditional gymnastics needed to maintain prevState and prevProps.

It will take time but I will look for specific tests that refute your concerns, and if I cannot find any I will write more tests.

Since you mentioned it, my concern now is what happens if `options.syncComponentUpdates` is `true` because there aren't any tests for that, and that is the only situation where what you describe might possibly happen, at least according to my admittedly limited understanding.